### PR TITLE
Fix cart items mismatch with swithcing guest/registered accounts

### PR DESCRIPTION
**Pull request by garmashdev**
# Description
Fixed sso cart items mismatch between customer / guest cards through the sso login process:
1. Added merging customer cart into guest cart and making guest cart as current active card for customers which is login via sso not in bolt modal (before this fix we are replacing customer cart by guest cart)
2. Added replacing customer cart by guest cart and making guest cart as current active card for customers which is login via sso in bolt modal. (before this fix we have situation when customer have 2 active carts after login in bolt pupup)

Fixes: https://app.asana.com/0/1200682492091884/1203162263301489/f

#changelog fix sso cart items mismatch between customer / guest cards after login.

# Type of change

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [x] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 

- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my ticket link and provided a changelog message.

### CODE REVIEW

File: Model/Api/OAuthRedirect.php [hunk 1]
```diff
@@ -351,10 +369,17 @@ public function login($code = '', $scope = '', $state = '', $reference = '')
                 // The checkout may not have been completed yet, but the user may have logged in via Bolt SSO
                 $quote = $this->cartHelper->getQuoteById($reference);
                 if ($quote !== false) {
+                    try {
+                        // we should make current customer cart as inactive and use guest cart as customer cart if login from bolt modal
+                        $customerActiveQuote = $this->cartRepository->getActiveForCustomer($customer->getId());
+                        $customerActiveQuote->setIsActive(false);
+                        $this->cartHelper->saveQuote($customerActiveQuote);
+                        // phpcs:ignore Magento2.CodeAnalysis.EmptyBlock
+                    } catch (\Magento\Framework\Exception\NoSuchEntityException $e) {
```
Comment (mikeshutov-bolt) [inline Model/Api/OAuthRedirect.php:378]: Can we at least log this exception it may help with debugging issues in the future? like a bugsnag log.

Comment (garmashdev) [reply Model/Api/OAuthRedirect.php:377]: Added bugsnag log exception.

**Comment by garmashdev** [comment]
> Did we test this with API driven and Plugin flow? because before this change it looked like Plugin flow worked.

I tested this solution on https://m2api.rc.dev.bolt.me/
Also i tested on https://m2.rc.dev.bolt.me/, but there we h
Снимок экрана 2022-11-17 в 17 40 15
ave some diff flow:

**Comment by mikeshutov-bolt** [comment]
> > Did we test this with API driven and Plugin flow? because before this change it looked like Plugin flow worked.
> 
> I tested this solution on https://m2api.rc.dev.bolt.me/ Also i tested on https://m2.rc.dev.bolt.me/, but there we have some diff flow: Снимок экрана 2022-11-17 в 17 40 15

You should be able to close this window (this is just our account window which probably opens on login)

**Comment by garmashdev** [comment]
> > > Did we test this with API driven and Plugin flow? because before this change it looked like Plugin flow worked.
> > 
> > 
> > I tested this solution on https://m2api.rc.dev.bolt.me/ Also i tested on https://m2.rc.dev.bolt.me/, but there we have some diff flow
> 
> You should be able to close this window (this is just our account window which probably opens on login)

Not, it is full width window, which is replacing normal account page.

File: Model/Api/OAuthRedirect.php [hunk 2]
```diff
@@ -365,11 +390,9 @@ public function login($code = '', $scope = '', $state = '', $reference = '')
                 $checkoutSession = $this->sessionHelper->getCheckoutSession();
                 if ($checkoutSession) {
                     $quote = $checkoutSession->getQuote();
-                    if ($quote) {
-                        $quote->setCustomer($customer);
-                        $quote->setCustomerIsGuest(false);
-                        $this->cartHelper->saveQuote($quote);
-
+                    if ($quote->getId()) {
+                        // we should merge guest customer cart with customer cart if login is not from bolt modal
```
Comment (vitaliyreznikov) [inline Model/Api/OAuthRedirect.php:394]: I guess `assignCustomer` is the function that merge two carts? If so, could we change the comment to "call the function that ..."

Comment (garmashdev) [reply Model/Api/OAuthRedirect.php]: comment was updated.

**Pull request merged**